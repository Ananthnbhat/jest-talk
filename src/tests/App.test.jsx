/* leave first line blank for cq */
import App from '../App';
import React from 'react';
import { shallow } from 'enzyme';

describe('App', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(
      <App />
    );
  });

  test('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot()
  })

  test('should have the `th` "Items"', () => {
    expect(
      wrapper.contains(<th>Items</th>)
    ).toBe(true);
  });

  test('should have a `button` element', () => {
    expect(
      wrapper.containsMatchingElement(
        <button>Add item</button>
      )
    ).toBe(true);
  });

  test('should have an `input` element', () => {
    expect(
      wrapper.containsMatchingElement(
        <input />
      )
    ).toBe(true);
  });

  test('`button` should be disabled', () => {
    const button = wrapper.find('button').first();
    expect(
      button.props().disabled
    ).toBe(true);
  });

  describe('the user populates the input', () => {
    const item = 'Vancouver';

    beforeEach(() => {
      const input = wrapper.find('input').first();
      input.simulate('change', {
        target: { value: item }
      });
    });

    test('should update the state property `item`', () => {
      expect(
        wrapper.state().item
      ).toEqual(item);
    });

    test('should enable `button`', () => {
      const button = wrapper.find('button').first();
      expect(
        button.props().disabled
      ).toBe(false);
    });

    describe('and then clears the input', () => {
      beforeEach(() => {
        const input = wrapper.find('input').first();
        input.simulate('change', {
          target: { value: '' }
        })
      });

      test('should disable `button`', () => {
        const button = wrapper.find('button').first();
        expect(
          button.props().disabled
        ).toBe(true);
      });
    });

    describe('and then submits the form', () => {
      beforeEach(() => {
        const form = wrapper.find('form').first();
        form.simulate('submit', {
          preventDefault: () => {},
        });
      });

      test('should add the item to state', () => {
        expect(
          wrapper.state().items
        ).toContain(item);
      });

      test('should render the item in the table', () => {
        expect(
          wrapper.containsMatchingElement(
            <td>{item}</td>
          )
        ).toBe(true);
      });

      test('should clear the input field', () => {
        const input = wrapper.find('input').first();
        expect(
          input.props().value
        ).toEqual('');
      });

      test('should disable `button`', () => {
        const button = wrapper.find('button').first();
        expect(
          button.props().disabled
        ).toBe(true);
      });
    });
  });
});
