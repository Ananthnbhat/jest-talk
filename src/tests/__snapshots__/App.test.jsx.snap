// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should match snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <App />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <table
        className="ui selectable structured large table"
>
        <thead>
                <tr>
                        <th>
                                Items
                        </th>
                </tr>
        </thead>
        <tbody>
                
        </tbody>
        <tfoot>
                <tr>
                        <th>
                                <form
                                        className="ui form"
                                        onSubmit={[Function]}
                                >
                                        <div
                                                className="field"
                                        >
                                                <input
                                                        className="prompt"
                                                        onChange={[Function]}
                                                        placeholder="Add item..."
                                                        type="text"
                                                        value=""
                                                />
                                        </div>
                                        <button
                                                className="ui button"
                                                disabled={true}
                                                type="submit"
                                        >
                                                Add item
                                        </button>
                                </form>
                        </th>
                </tr>
        </tfoot>
</table>,
      "className": "ui text container",
      "id": "app",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <thead>
            <tr>
                        <th>
                                    Items
                        </th>
            </tr>
</thead>,
          <tbody>
            
</tbody>,
          <tfoot>
            <tr>
                        <th>
                                    <form
                                                className="ui form"
                                                onSubmit={[Function]}
                                    >
                                                <div
                                                            className="field"
                                                >
                                                            <input
                                                                        className="prompt"
                                                                        onChange={[Function]}
                                                                        placeholder="Add item..."
                                                                        type="text"
                                                                        value=""
                                                            />
                                                </div>
                                                <button
                                                            className="ui button"
                                                            disabled={true}
                                                            type="submit"
                                                >
                                                            Add item
                                                </button>
                                    </form>
                        </th>
            </tr>
</tfoot>,
        ],
        "className": "ui selectable structured large table",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <tr>
              <th>
                            Items
              </th>
</tr>,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <th>
                Items
</th>,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Items",
              },
              "ref": null,
              "rendered": "Items",
              "type": "th",
            },
            "type": "tr",
          },
          "type": "thead",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [],
          },
          "ref": null,
          "rendered": Array [],
          "type": "tbody",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <tr>
              <th>
                            <form
                                          className="ui form"
                                          onSubmit={[Function]}
                            >
                                          <div
                                                        className="field"
                                          >
                                                        <input
                                                                      className="prompt"
                                                                      onChange={[Function]}
                                                                      placeholder="Add item..."
                                                                      type="text"
                                                                      value=""
                                                        />
                                          </div>
                                          <button
                                                        className="ui button"
                                                        disabled={true}
                                                        type="submit"
                                          >
                                                        Add item
                                          </button>
                            </form>
              </th>
</tr>,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <th>
                <form
                                className="ui form"
                                onSubmit={[Function]}
                >
                                <div
                                                className="field"
                                >
                                                <input
                                                                className="prompt"
                                                                onChange={[Function]}
                                                                placeholder="Add item..."
                                                                type="text"
                                                                value=""
                                                />
                                </div>
                                <button
                                                className="ui button"
                                                disabled={true}
                                                type="submit"
                                >
                                                Add item
                                </button>
                </form>
</th>,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <form
                  className="ui form"
                  onSubmit={[Function]}
>
                  <div
                                    className="field"
                  >
                                    <input
                                                      className="prompt"
                                                      onChange={[Function]}
                                                      placeholder="Add item..."
                                                      type="text"
                                                      value=""
                                    />
                  </div>
                  <button
                                    className="ui button"
                                    disabled={true}
                                    type="submit"
                  >
                                    Add item
                  </button>
</form>,
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": Array [
                    <div
                      className="field"
>
                      <input
                                            className="prompt"
                                            onChange={[Function]}
                                            placeholder="Add item..."
                                            type="text"
                                            value=""
                      />
</div>,
                    <button
                      className="ui button"
                      disabled={true}
                      type="submit"
>
                      Add item
</button>,
                  ],
                  "className": "ui form",
                  "onSubmit": [Function],
                },
                "ref": null,
                "rendered": Array [
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "host",
                    "props": Object {
                      "children": <input
                        className="prompt"
                        onChange={[Function]}
                        placeholder="Add item..."
                        type="text"
                        value=""
/>,
                      "className": "field",
                    },
                    "ref": null,
                    "rendered": Object {
                      "instance": null,
                      "key": undefined,
                      "nodeType": "host",
                      "props": Object {
                        "className": "prompt",
                        "onChange": [Function],
                        "placeholder": "Add item...",
                        "type": "text",
                        "value": "",
                      },
                      "ref": null,
                      "rendered": null,
                      "type": "input",
                    },
                    "type": "div",
                  },
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "host",
                    "props": Object {
                      "children": "Add item",
                      "className": "ui button",
                      "disabled": true,
                      "type": "submit",
                    },
                    "ref": null,
                    "rendered": "Add item",
                    "type": "button",
                  },
                ],
                "type": "form",
              },
              "type": "th",
            },
            "type": "tr",
          },
          "type": "tfoot",
        },
      ],
      "type": "table",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <table
          className="ui selectable structured large table"
>
          <thead>
                    <tr>
                              <th>
                                        Items
                              </th>
                    </tr>
          </thead>
          <tbody>
                    
          </tbody>
          <tfoot>
                    <tr>
                              <th>
                                        <form
                                                  className="ui form"
                                                  onSubmit={[Function]}
                                        >
                                                  <div
                                                            className="field"
                                                  >
                                                            <input
                                                                      className="prompt"
                                                                      onChange={[Function]}
                                                                      placeholder="Add item..."
                                                                      type="text"
                                                                      value=""
                                                            />
                                                  </div>
                                                  <button
                                                            className="ui button"
                                                            disabled={true}
                                                            type="submit"
                                                  >
                                                            Add item
                                                  </button>
                                        </form>
                              </th>
                    </tr>
          </tfoot>
</table>,
        "className": "ui text container",
        "id": "app",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <thead>
              <tr>
                            <th>
                                          Items
                            </th>
              </tr>
</thead>,
            <tbody>
              
</tbody>,
            <tfoot>
              <tr>
                            <th>
                                          <form
                                                        className="ui form"
                                                        onSubmit={[Function]}
                                          >
                                                        <div
                                                                      className="field"
                                                        >
                                                                      <input
                                                                                    className="prompt"
                                                                                    onChange={[Function]}
                                                                                    placeholder="Add item..."
                                                                                    type="text"
                                                                                    value=""
                                                                      />
                                                        </div>
                                                        <button
                                                                      className="ui button"
                                                                      disabled={true}
                                                                      type="submit"
                                                        >
                                                                      Add item
                                                        </button>
                                          </form>
                            </th>
              </tr>
</tfoot>,
          ],
          "className": "ui selectable structured large table",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <tr>
                <th>
                                Items
                </th>
</tr>,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <th>
                  Items
</th>,
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Items",
                },
                "ref": null,
                "rendered": "Items",
                "type": "th",
              },
              "type": "tr",
            },
            "type": "thead",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [],
            },
            "ref": null,
            "rendered": Array [],
            "type": "tbody",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <tr>
                <th>
                                <form
                                                className="ui form"
                                                onSubmit={[Function]}
                                >
                                                <div
                                                                className="field"
                                                >
                                                                <input
                                                                                className="prompt"
                                                                                onChange={[Function]}
                                                                                placeholder="Add item..."
                                                                                type="text"
                                                                                value=""
                                                                />
                                                </div>
                                                <button
                                                                className="ui button"
                                                                disabled={true}
                                                                type="submit"
                                                >
                                                                Add item
                                                </button>
                                </form>
                </th>
</tr>,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <th>
                  <form
                                    className="ui form"
                                    onSubmit={[Function]}
                  >
                                    <div
                                                      className="field"
                                    >
                                                      <input
                                                                        className="prompt"
                                                                        onChange={[Function]}
                                                                        placeholder="Add item..."
                                                                        type="text"
                                                                        value=""
                                                      />
                                    </div>
                                    <button
                                                      className="ui button"
                                                      disabled={true}
                                                      type="submit"
                                    >
                                                      Add item
                                    </button>
                  </form>
</th>,
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": <form
                    className="ui form"
                    onSubmit={[Function]}
>
                    <div
                                        className="field"
                    >
                                        <input
                                                            className="prompt"
                                                            onChange={[Function]}
                                                            placeholder="Add item..."
                                                            type="text"
                                                            value=""
                                        />
                    </div>
                    <button
                                        className="ui button"
                                        disabled={true}
                                        type="submit"
                    >
                                        Add item
                    </button>
</form>,
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": Array [
                      <div
                        className="field"
>
                        <input
                                                className="prompt"
                                                onChange={[Function]}
                                                placeholder="Add item..."
                                                type="text"
                                                value=""
                        />
</div>,
                      <button
                        className="ui button"
                        disabled={true}
                        type="submit"
>
                        Add item
</button>,
                    ],
                    "className": "ui form",
                    "onSubmit": [Function],
                  },
                  "ref": null,
                  "rendered": Array [
                    Object {
                      "instance": null,
                      "key": undefined,
                      "nodeType": "host",
                      "props": Object {
                        "children": <input
                          className="prompt"
                          onChange={[Function]}
                          placeholder="Add item..."
                          type="text"
                          value=""
/>,
                        "className": "field",
                      },
                      "ref": null,
                      "rendered": Object {
                        "instance": null,
                        "key": undefined,
                        "nodeType": "host",
                        "props": Object {
                          "className": "prompt",
                          "onChange": [Function],
                          "placeholder": "Add item...",
                          "type": "text",
                          "value": "",
                        },
                        "ref": null,
                        "rendered": null,
                        "type": "input",
                      },
                      "type": "div",
                    },
                    Object {
                      "instance": null,
                      "key": undefined,
                      "nodeType": "host",
                      "props": Object {
                        "children": "Add item",
                        "className": "ui button",
                        "disabled": true,
                        "type": "submit",
                      },
                      "ref": null,
                      "rendered": "Add item",
                      "type": "button",
                    },
                  ],
                  "type": "form",
                },
                "type": "th",
              },
              "type": "tr",
            },
            "type": "tfoot",
          },
        ],
        "type": "table",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
